[libc++][memory] P2652R2: Disallow Specialization of `allocator_traits`

H-G-Hristov opened the pull request:
Implements P2652R2 <https://wg21.link/P2652R2>:
- https://eel.is/c++draft/allocator.requirements.general
- https://eel.is/c++draft/memory.syn
- https://eel.is/c++draft/allocator.traits.general
- https://eel.is/c++draft/allocator.traits.members
- https://eel.is/c++draft/diff.cpp20.concepts
- https://eel.is/c++draft/diff.cpp20.utilities

Zingam committed:
[libc++][memory] P2652R2 - Disallow Specialization of `allocator_traits`

Implements P2652R2 <https://wg21.link/P2652R2>:
- https://eel.is/c++draft/allocator.requirements.general
- https://eel.is/c++draft/memory.syn
- https://eel.is/c++draft/allocator.traits.general
- https://eel.is/c++draft/allocator.traits.members
- https://eel.is/c++draft/diff.cpp20.concepts
- https://eel.is/c++draft/diff.cpp20.utilities

philnik777 commented on libcxx/docs/ReleaseNotes/19.rst:
```diff
@@ -37,12 +37,16 @@ What's New in Libc++ 19.0.0?
 -----------------------------
-TODO
+
+- The ``_LIBCPP_ENABLE_CXX23_USER_SPECIALIZATION_OF_ALLOCATOR_TRAITS`` macro has been added to re-enable user specialization
```
There is no need to add a flag for this. The feature has been added in C++23, and not many users will have started relying on it.

H-G-Hristov replied:
I thought so, removed it. Thank you for reviewing!

philnik777 reviewed (changes requested):
Other than removing the flag this looks pretty good.

H-G-Hristov committed:
Removed redundant fllag: `_LIBCPP_ENABLE_CXX23_USER_SPECIALIZATION_OF_ALLOCATOR_TRAITS`

Since this features was introduced in C++23 there is no need for a re-enable flag and the redundant code was removed.

H-G-Hristov committed:
Merge branch 'main' into hgh/libcxx/P2652R2-Disallow-User-Specialization-of-allocator_traits

H-G-Hristov committed:
Fixed formatting

H-G-Hristov committed:
Merge branch 'main' into hgh/libcxx/P2652R2-Disallow-User-Specialization-of-allocator_traits

philnik777 commented on libcxx/utils/generate_feature_test_macro_components.py:
```diff
@@ -88,9 +88,9 @@ def add_version_header(tc):
             "name": "__cpp_lib_allocate_at_least",
             "values": {
-                "c++23": 202106,
+                # "c++23": 202106,  # P0401R6 Providing size feedback in the Allocator interface
```
This line isn't needed anymore.

Zingam replied:
```suggestion
```

Zingam commented on libcxx/include/__memory/allocation_result.h:
```diff
@@ -0,0 +1,34 @@
+
+#if _LIBCPP_STD_VER >= 23
+
+template <class _Pointer, class _SizeType = size_t>
```
To unbreak a circular dependancy.

H-G-Hristov replied:
I guess this can be folded into `allocator_traits.h`.

Zingam committed:
Update libcxx/utils/generate_feature_test_macro_components.py

Zingam committed:
Merge branch 'main' into hgh/libcxx/P2652R2-Disallow-User-Specialization-of-allocator_traits

H-G-Hristov committed:
Cleanup

H-G-Hristov committed:
Removed `allocation_result.h`

H-G-Hristov committed:
Merge branch 'main' into hgh/libcxx/P2652R2-Disallow-User-Specialization-of-allocator_traits

H-G-Hristov committed:
Undo formatting of <string>

H-G-Hristov committed:
Fix CI

H-G-Hristov committed:
Merge branch 'main' into hgh/libcxx/P2652R2-Disallow-User-Specialization-of-allocator_traits

H-G-Hristov committed:
Fixed syn

philnik777 reviewed (approved):
Thanks! LGTM.
